src/tools/rbd_mirror/image_replayer/snapshot/Replayer.cc

```diff
@@ -898,74 +930,134 @@ void Replayer<I>::handle_unlink_peer(int r) {
     notify_status_updated();
   }
 
-  refresh_local_image();
+  load_local_image_meta();
 }
 
 template <typename I>
-void Replayer<I>::register_update_watcher() {
+void Replayer<I>::register_local_update_watcher() {
   dout(10) << dendl;
 
   m_update_watch_ctx = new C_UpdateWatchCtx(this);
```

Comment:
@dillaman The memcheck reports error:
```
2020-02-27T00:56:54.187 INFO:tasks.workunit.client.0.smithi061.stderr:==28758== 352 bytes in 22 blocks are definitely lost in loss record 7 of 11
2020-02-27T00:56:54.187 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    at 0x4C31556: operator new(unsigned long) (vg_replace_malloc.c:344)
2020-02-27T00:56:54.187 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x56CC80: rbd::mirror::image_replayer::snapshot::Replayer<librbd::ImageCtx>::register_local_update_watcher() (Replayer.cc:940)
2020-02-27T00:56:54.187 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x56D224: rbd::mirror::image_replayer::snapshot::Replayer<librbd::ImageCtx>::init(Context*) (Replayer.cc:157)
2020-02-27T00:56:54.187 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x492F34: rbd::mirror::ImageReplayer<librbd::ImageCtx>::start_replay() (ImageReplayer.cc:417)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x4931D4: rbd::mirror::ImageReplayer<librbd::ImageCtx>::handle_bootstrap(int) (ImageReplayer.cc:402)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x49B6FF: complete (Context.h:77)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x49B6FF: rbd::mirror::CancelableRequest::finish(int) (CancelableRequest.h:30)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x512216: rbd::mirror::image_replayer::BootstrapRequest<librbd::ImageCtx>::close_remote_image() (BootstrapRequest.cc:422)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x513A54: rbd::mirror::image_replayer::BootstrapRequest<librbd::ImageCtx>::handle_prepare_replay(int) (BootstrapRequest.cc:320)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x565843: complete (Context.h:77)
2020-02-27T00:56:54.188 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x565843: finish (BaseRequest.h:22)
2020-02-27T00:56:54.189 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x565843: rbd::mirror::image_replayer::snapshot::PrepareReplayRequest<librbd::ImageCtx>::handle_load_local_image_meta(int) (PrepareReplayRequest.cc:62)
2020-02-27T00:56:54.189 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x7E25F7: complete (Context.h:77)
2020-02-27T00:56:54.189 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x7E25F7: librbd::mirror::snapshot::ImageMeta<librbd::ImageCtx>::handle_load(Context*, int) (ImageMeta.cc:71)
2020-02-27T00:56:54.189 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x7DF99B: complete (Context.h:77)
2020-02-27T00:56:54.189 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x7DF99B: void librbd::util::detail::rados_callback<LambdaContext<librbd::mirror::snapshot::ImageMeta<librbd::ImageCtx>::load(Context*)::{lambda(int)#2}> >(void*, LambdaContext<librbd::mirror::snapshot::ImageMeta<librbd::ImageCtx>::load(Context*)::{lambda(int)#2}>) (Utils.h:27)
2020-02-27T00:56:54.189 INFO:tasks.workunit.client.0.smithi061.stderr:==28758==    by 0x4EE02A0: librados::C_AioComplete::finish(int) (AioCompletionImpl.h:140)
```
We use this variable for both local and remote watch contexts and delete only once.

[1] http://qa-proxy.ceph.com/teuthology/trociny-2020-02-26_17:11:11-rbd-wip-43934-distro-basic-smithi/4805442/teuthology.log

Reply:
Oops -- definitely only need a single update watch context.

Reply:
Deleted the second `new` operation in `register_remote_update_watcher` below.